### DISCUSSION TITLE
Postgres POSIX regular expressions

### DISCUSSION THREAD

**Discussion by HypeillingerChillinger** [linq2db/linq2db#4183] [Q&A]
Is there a built-in way to use POSIX regular expressions?

If not how can this be used SQL injection safe?

```csharp
[Sql.Expression(ProviderName.PostgreSQL, "{0} ~* {1}", ServerSideOnly = true)]
public static bool Custom(string data, string regEx)
{
     throw new NotImplementedException();
}
```

**Comment by jods4** [comment]
If this is used with variables, linq2db would use SQL parameters and so it's injection-safe, isn't it?

If you pass a literal string, it's not injection-safe but as it's a literal string in your code you'd be in full control.

**Reply by HypeillingerChillinger** [reply]
I pass a string literal. So in other words, there is no integrated method in Linq2Db?

Is there a method in Linq2Db that I could use to safely escape special chars in the literal?

**Reply by jods4** [reply]
If it's a literal, it's a constant in your code? What is the injection vector here?

In general there's no way a downstream consumer such as linq2db could be aware of what tokens are part of the intended syntax and what tokens should be escaped.
If some input comes from users and needs to be escaped, you should do it at the application-level, with the knowledge of what that input is supposed to contain.

**Reply by HypeillingerChillinger** [reply]
Ok, thanks

**Comment by HypeillingerChillinger** [comment]
It seems that Linq2Db already escapes - at least single quotes. So is it already safe to use?
```csharp
.Where(x => MyExpressions.Custom(x.BB, "^asdf.*sdf$' OR 1=1 '"))

```
```sql 
WHERE x.bla ~* '^asdf.*sdf$'' OR 1=1 ''' = True AND x.bla ~* '^asdf.*sdf$'' OR 1=1 ''' IS NOT NULL
```

**Reply by jods4** [reply]
linq2db knows how to write/escape any string value such as `abc'123` into a SQL literal.

SQL strings are single-quoted, and `'` that appears inside it are doubled. That level of escaping (SQL string literal representation) is easy and linq2db can generate `'abc''123'`.

That's where it stops. It wouldn't know anything about the regex content, for example is `user-.*? | .*` an attempt at bypassing a filter or legitimate? (Also note that regex escaping depends on the regex dialect and they're not the same on every database.)

> So is it already safe to use?

To answer this question, you'd need to precisely define "safe"? 

> **Note**
> Maybe not directly related to your question, but note that injecting uncontrolled strings is a tricky business. There are many attack vectors. 
> Injection can use length-limitation to truncate criteria off of a request. 
> Some DB accept `/*` and `*/` inside strings to comment away multiple restrictions between two injected strings. 
> See for example: https://www.directdefense.com/limited-length-sql-injection/

**Comment by HypeillingerChillinger** [comment]
Ok, the definition of safe in my case would be to be protected against SQL injection apart from manipulating the RegEx in order to extend the number of matches like in your example. 

Can it then be considered to be safe?

**Reply by jods4** [reply]
linq2db does properly write string literals in generated SQL, i.e. quotes are escaped (doubled), so it "works".

As illustrated by the link I gave you above, this may not always be sufficient. Sometimes more complex attacks are possible, e.g. based on overflowing the length of your query.

When it comes to security, the common advice is to always put user-controlled strings into SQL parameters, never inline them into a query.

**Reply by HypeillingerChillinger** [reply]
Ok, thanks.
